src/Analyzers/CSharp/CodeFixes/RemoveUnnecessaryNullableDirective/CSharpRemoveUnnecessaryNullableDirectiveCodeFixProvider.cs

```diff
@@ -53,10 +57,46 @@ public override Task RegisterCodeFixesAsync(CodeFixContext context)
             CodeActionOptionsProvider fallbackOptions,
             CancellationToken cancellationToken)
         {
-            foreach (var diagnostic in diagnostics)
+            var nullableDirectivesByNodes = diagnostics.GroupBy(x =>
+                x.Location.FindNode(findInsideTrivia: false, getInnermostNodeForTie: false, cancellationToken), x =>
+                x.Location.FindNode(findInsideTrivia: true, getInnermostNodeForTie: true, cancellationToken))
+                .ToDictionary(x => x.Key, x => x.ToList());
+
+            foreach (var (node, nullableDirectives) in nullableDirectivesByNodes)
             {
-                var nullableDirective = diagnostic.Location.FindNode(findInsideTrivia: true, getInnermostNodeForTie: true, cancellationToken);
-                editor.RemoveNode(nullableDirective, SyntaxRemoveOptions.KeepNoTrivia);
+                if (node is CompilationUnitSyntax)
+                {
+                    foreach (var nullableDirective in nullableDirectives)
+                        editor.RemoveNode(nullableDirective, SyntaxRemoveOptions.KeepNoTrivia);
+                    continue;
+                }
+
+                var leadingTrivia = node.GetLeadingTrivia();
+                var indexes = nullableDirectives.Select(x => leadingTrivia.IndexOf(x.ParentTrivia))
+                                                .OrderByDescending(x => x).ToArray();
+                foreach (var index in indexes)
+                {
+                    var i = index;
+                    leadingTrivia = leadingTrivia.RemoveAt(i);
+                    i--;
+                    while (i >= 0)
+                    {
+                        if (leadingTrivia[i].IsEndOfLine())
+                        {
+                            leadingTrivia = leadingTrivia.RemoveAt(i);
+                            break;
+                        }
+                        if (leadingTrivia[i].IsWhitespace())
+                        {
+                            leadingTrivia = leadingTrivia.RemoveAt(i);
+                        }
+                        i--;
+                    }
+                }
```

Comment:
it's unclear why it's not sufficient to just remove all the trivia that precedes the nullable directive that is whitespace or endofline.

Reply:
It would remove more lines than desired if you removed all of them like if we had 2 blank lines before. This is why I remove at most one preceding line

Reply:
so, basically, i see this as having a much simpler solution.

check the directive.  we're looking to see if there's a blank line above it, and a blank line below it.

if there is both, then we want to simply remove one of those blank lines.  that seems to be it afaict.  if there are no blank lines we don't need to do anything.  and if htere is just a blank line above *or* a blank line below (but not both) we don't need to do anything either.

the code seems very overwrought to accomplish this.

Reply:
I agree it seemed weird to me, but I couldn't find an easier to get the preceding trivia or the following one.

Maybe there was an api or method of traversal I wasn't aware of?